### ISSUE TITLE
ArcSpellLvlMod DivSpellLvlMod (Pale Master) do not correctly add Caster Levels and Learned Spells

### ISSUE THREAD

**Issue by Finaldeath** [Beamdog/nwn-issues#207]
**To Reproduce**

Caster Level should, according to D&D (and how NWN2 does it) scale the Caster Level by the same amount as the ArcSpellLvlMod does increase levels for the case of spell slots. Without this caster levels are incredibly low for instance:

 - Load this test debug module with the pre-generated character "Undead Lord" (Fighter 4/ Wizard 6 / Pale Master 30): 
[Jasperre Debug Spells Cast.zip](https://github.com/Beamdog/nwn-issues/files/5469408/Jasperre.Debug.Spells.Cast.zip)
- Cast a spell
- Check the "Caster Level". Reports as 6 instead of what it should be - 6 + (30/2 = 15)  = 21

The game should also grant spells learnt on level up - eg; Wizards get 2 new ones, and Sorcerers and Bards would get new spells to swap in/out. Without this effectively the column ArcSpellLvlMod is unusable for any Bards and Sorcerers. To see in action:

- Create Sorcerer who is not good
- Debug add 1000000 xp in console
- Levelup to get level 3 spells in Sorcerer
- Take Pale Master from now on - no new spells area available. Your spell book is essentially blank - for instance you eventually get access to level 4 spells...with nothing available to put in them

**Specifics**

 - OS Windows
 - Game build 82.8193.18 [6f8f5b92]
 - Distro platform Steam
 - Content overrides? None
 - Regression? Always been like this

**If needed, describe the bug**

Noted above the issues. The solution would be to have the Pale Master logically "latch on" to a single class (perhaps the first class in numeric order that is "Arcane" in the case of ArcSpellLvlMod) and grant the same level up learned spells and caster level to that class as the spell class gets spell slots.

Without this as seen in most of the uses in the game you get it being used only for melee classes or to get critical immunity. Hardly ever do actual spellcasters use it because:
* You cannot penetrate good Spell Resistance
* Spells which are in rounds/level last only a few rounds before going
* As a Sorcerer and Bard you gain such a penalty - no new spells to learn - the slots you do gain eventually go empty

The nwscript functions that would need updates: GetCasterLevel, ResistSpell and ApplyEffect functions (that store the caster level for future dispel attempts and for AOEs). The game would need some ELC changes as well of course.

For modders anyone using these two columns for other prestige classes also suffer these penalties making them pretty useless for actual spellcasters and anyone who doesn't learn spells from scrolls.

This was fixed in NWN2; so, hopefully fixable in NWN1 :)

**Comment by pF-arQon** [comment]
No, "GetCasterLevel, ResistSpell and ApplyEffect functions ..." need updating because they're don't work correctly in the first place, and never have.
EMS had to work around those bugs by rewriting literally every spell script in the game, which happened to also make supporting CL stacking for PM (and custom Arcane PRCs) trivial - but that also has hundreds of balance changes, so it's not for everyone.

The "spells known" issue though has no possible "userspace" workarounds, and as you say pretty much "breaks" spontaneous casters. Given that, if it comes to it I'd much rather see this split into two separate issues, with the "spells known" aspect carrying a much higher priority than the CL stacking.

**Comment by Stratovarius667** [comment]
Will explicitly support this issue - we at the PRC have had to do a significant amount of custom coding to overcome the problems detailed here, and it would be excellent were they fixed.

> The "spells known" issue though has no possible "userspace" workarounds, and as you say pretty much "breaks" spontaneous casters. Given that, if it comes to it I'd much rather see this split into two separate issues, with the "spells known" aspect carrying a much higher priority than the CL stacking.

Incorrect. PRC has a userspace workaround. This does not detract from the severity of the bug, and the fact we'd like to see it fixed as much as anyone else would. Would make our lives much easier.
